Upload JSON feature

Cthomp7 opened the pull request:
This is what I created for the upload JSON button for the Demo/Example Maker. (additionally linted two other lines in the file)

If there are better ways to structure things (it's a bit long), I would love to make this as efficient and condensed as I can.

I want to add a "do you want to upload this file?" convo message before loading the JSON content and need to fix a toc issue, but wanted to start the PR to discuss revisions of what I have already created before moving forward. :)

nbriz commented on www/widgets/DemoExampleMaker.js:
```diff
@@ -291,6 +298,60 @@ class DemoExampleMaker extends Widget {
   }
 
+  _uploadJSON (dem) {
+    const fileInput = document.createElement('input')
+    fileInput.type = 'file'
+    fileInput.addEventListener('change', function () {
```
hey, let's try to use our internal `nn.FileUploader` class to keep the code a little tighter (instead of rolling this out from scratch). u can find documentation for that [here](https://github.com/netizenorg/netnet-standard-library/blob/main/docs/API.md#FileUploader)

nbriz commented on www/widgets/DemoExampleMaker.js:
```diff
@@ -291,6 +298,60 @@ class DemoExampleMaker extends Widget {
     this.$('[name="dem-url"]').style.display = 'block'
   }
 
+  _uploadJSON (dem) {
```
lets use arrow function syntax `() => { }` so we can access `this` of the class && avoid having to pass this param in here

nbriz reviewed (commented):
double check to see if `toc` isn't actually working, if so can u open up an Issue (maybe also include in there what we discussed about changing the value "true/false" instead of "on/off")

also confirm that `.info` can't be mapped directly to the `steps` property (if not then loop through the steps like u're already doing here)

Cthomp7 commented:
Hey @nbriz! I have fixed the new code according to the comments above.

To confirm:
- toc was not being recorded as 'off' when clicked off and saved. (will make issue)
- `.info` can't be mapped directly to the `steps` property

I was receiving standard JS errors for the `const fileUploader = new nn.FileUploader({...` , so I added the `Object.create(fileUploader)` to fix this. I am not sure if this is the _proper way_ to do this, so that's why I wanted to point this out and get your feedback. Thanks!

nbriz commented on www/widgets/DemoExampleMaker.js:
```diff
@@ -291,6 +310,47 @@ class DemoExampleMaker extends Widget {
+    NNE.code = NNE._decode(ex.code.split('#code/').pop())
+    if (ex.info) {
+      this._data.steps.forEach((step, i) => {
+        if (i === 0 || i < this._data.steps.length) {
```
this is looking great @Cthomp7, i've got some notes but b4 i post the full review, could u help me understand why this conditional statement is here?

nbriz replied:
there might be something i'm missing (let me know if i'm over looking a case here), but i'm thinking this should look more like:

```js
    if (ex.info) {
      this._data.steps.forEach((step, i) => {
        this._addStep(step)
        this._selectStep(0)
      })
    } else { 
      this.$('[name="dem-s-title"]').value = null
      this.$('[name="dem-s-focus"]').value = null
      this.$('[name="dem-s-text"]').value = null
    }
```
so that if the example does have "info" (ie. steps) than we add each one && select the first one by default, otherwise we keep the steps array empty && null out the view fields. what do u think?

Cthomp7 replied:
I had line 335 above since before I rewriting over already existing steps and since there always has to be one step I was always updating the preexisting first step. **However** the `steps: ex.info || [{}]` now works for creating the steps in the widget, so I have removed that section of the code in my newest revision.

nbriz commented on www/widgets/DemoExampleMaker.js:
```diff
@@ -291,6 +310,47 @@ class DemoExampleMaker extends Widget {
+      code: ex.code,
+      steps: ex.info ||
+      [{ title: 'getting started', focus: null, text: 'in this step...' }]
+    }
+    console.log(ex.toc)
+    this.$('[name="dem-demo-name"]').value = ex.name
```
let's remove this `console.log()`

also, though the default "dock-left" layout is likely to be up most of the time, i think it could be helpful to check what layout netnet is in here && adjust accordingly (tiny "edge case" but as i was testing sometimes netnet was in the "welcome" layout which means i couldn't see the code after uploading it), so something like

```js
    if (NNW.layout !== this._data.layout) {
      NNW.layout = this._data.layout
    }
```

nbriz commented on www/widgets/DemoExampleMaker.js:
```diff
@@ -291,6 +310,47 @@ class DemoExampleMaker extends Widget {
+    this._data = {
+      name: ex.name,
+      tags: ex.tags || null,
+      toc: ex.toc || true,
```
if the example doesn't have a `toc` than that means it's likely not annotated, so while i think the default value for a newly created example with annotations should be `true`, in this case (the default value for an old example we're uploading that's missing a `toc`) should be `false` (otherwise it opens the toc even though there aren't any steps to really go through)

nbriz commented on www/widgets/DemoExampleMaker.js:
```diff
@@ -291,6 +310,47 @@ class DemoExampleMaker extends Widget {
+      key: Number(ex.key),
+      code: ex.code,
+      steps: ex.info ||
+      [{ title: 'getting started', focus: null, text: 'in this step...' }]
```
similar to my note above, if we're opening an old example that doesn't have any steps, we shouldn't create one (we should let them create it) so the default value here should just be an empty array

nbriz reviewed (changes requested):
thnx for refactoring this! did some more testing && noticed a few things (see comments below)

Cthomp7 commented:
@nbriz, I have made a new commit with the changes requested above.

Let me know if there are any other issues!
